dspace-api: tell ImageMagick about the PDF CropBox

alanorth opened the pull request:
## References
* Fixes: https://github.com/DSpace/DSpace/issues/8549

## Description
ImageMagick uses the `MediaBox` by default when rasterizing PDFs because the PDF specification says that all PDFs *must* contain one. This page box is the parent for all other boxes that a PDF *may* contain, for example a `CropBox`, `ArtBox`, etc. In many cases these are the same, but when they are not the `CropBox` is generally used to define the area displayed to a user when they open the PDF on screen (as opposed to when printing on paper). Using the wrong page box can result in strange-looking thumbnails.

If a PDF has a `CropBox` that has a different size than its `MediaBox` then we should tell ImageMagick to use it because it more accurately resembles what the user would see if they opened the PDF.

**Before (open the image to see the blank white space):**

<p align="center">
  <img width="400" src="https://user-images.githubusercontent.com/191754/198273645-198155ca-1765-47fa-8887-53cbcf2cca6d.jpg">
</p>

**After:**

<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/191754/198274172-21e8b6d6-1730-4b45-b979-091bddc41613.jpg">
</p>

And example PDF can be found in our repository: https://hdl.handle.net/10568/116598

## Instructions for Reviewers
Test generating an ImageMagick PDF thumbnail for several items before and after this patch, ie:

```console
$ dspace filter-media -p 'ImageMagick PDF Thumbnail' -v -f -i 10568/120157
```

List of changes in this PR:
* Include several imports from pdfbox (which we are already using elsewhere) in `dspace-api/src/main/java/org/dspace/app/mediafilter/ImageMagickThumbnailFilter.java`
* Add logic to check the `MediaBox` and `CropBox`
* Add logic to pass the `-define pdf:use-cropbox=true` option to ImageMagick

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).

alanorth commented on dspace-api/src/main/java/org/dspace/app/mediafilter/ImageMagickThumbnailFilter.java:
```diff
@@ -119,6 +122,25 @@ public File getImageFile(File f, int page, boolean verbose)
+        // CropBox. Note: we don't need to do anything special to detect if
+        // the CropBox is missing or empty because pdfbox will set it to the
+        // same size as the MediaBox if it doesn't exist.
+        PDPage pdfPage = PDDocument.load(f).getPage(1);
```
We only need to test page one since we are only generating a thumbnail from page one. I'm not sure if we need to add that to the inline comment?

alanorth replied:
There was actually an issue here in that PDDocument's `getPage()` uses a zero-based index, so I was actually checking the *second* page and this was failing on PDFs with only one page!

nwoodward commented:
Do you have an example PDF in your repository I could use to test this?

alanorth commented:
> Do you have an example PDF in your repository I could use to test this?

Ah yes, I should have linked one! Here is the sample I used: https://hdl.handle.net/10568/116598

nwoodward reviewed (approved):
👍 tested and confirmed this works. 

I looked but couldn't find when the use-cropbox option was added to ImageMagick or if it's been there forever. There isn't a minimum version of ImageMagick mentioned in the DSpace 7 documentation, but if the cropbox option isn't available in older versions then it would be a good idea to note that.

alanorth commented:
> I looked but couldn't find when the use-cropbox option was added to ImageMagick or if it's been there forever.

Good question! Using some rudimentary `git show` I checked the ImageMagick history for `coders/pdf.c` and found the option existing in [git commits dating back to at least 2015](https://github.com/ImageMagick/ImageMagick/blob/151b66dffc9e3c2e8c4f8cdaca37ff987ca0f497/coders/pdf.c#L508).

Also, I did a test on a CentOS 7 box I still have laying around and the `-define pdf:use-cropbox=true` option works correctly with ImageMagick 6.9.10.68 there, so I think it's pretty safe to assume this will work "everywhere" for current systems.

alanorth committed:
dspace-api: tell ImageMagick about the PDF CropBox

ImageMagick uses the MediaBox by default when rasterizing PDFs be-
cause the PDF specification says that all PDFs *must* contain one.
This page box is the parent for all other boxes that a PDF *may*
contain, for example a CropBox, ArtBox, etc. In many cases these
are the same, but when they are not the CropBox is used to define
the area displayed to a user when they open the PDF on screen (as
opposed to when printing on paper).

If a PDF has a CropBox that is different to its MediaBox then we
should tell ImageMagick to use it.

Fixes: https://github.com/DSpace/DSpace/issues/8549

alanorth commented:
I found an issue with my original patch: PDDocument's `getPage()` is actually zero based, so we need to get page 0 if we want to check the first page. I've updated the patch.

tdonohue commented:
@alanorth : I've used the automatic rebase button on this PR as it was previously failing on a broken test (and that broken test has been fixed on `main` now).

alanorth commented:
@tdonohue that's OK. It's only one commit so it is clean and easy to rebase. Thanks.

tdonohue reviewed (approved):
👍 Thanks @alanorth !  This looks good to me.  I tested it today and verified the sample PDF creates a correct thumbnail.  I also tested it against a few other PDFs just to be certain there are no other side effects -- and I didn't find any.
